### ISSUE TITLE
Closing an async DatalakeFileClient (from get_file_client) makes subsequent DatalakeFileClient operations fail

### ISSUE THREAD

**Issue by oliverw1** [Azure/azure-sdk-for-python#29407]
- **Package Name**:  azure-storage-filedatalake
- **Package Version**: 12.9.1, 12.10.1
- **Operating System**: Linux
- **Python Version**: 3.10.10

**Describe the bug**
Similar issue as [#14497](https://github.com/Azure/azure-sdk-for-python/issues/14497):
> Opening a FileSystemClient as a context manager and then getting multiple file clients, if the first one is closed, all subsequent DataLakeFileClient operations will fail.

**To Reproduce**

```python
import asyncio
from azure.storage.filedatalake.aio import DataLakeDirectoryClient

conn_str = "REDACTED"

async def main():
    async with DataLakeDirectoryClient.from_connection_string(
        conn_str, file_system_name="foo", directory_name="bar"
    ) as dir_client:
            async with dir_client.get_file_client("foo1") as file_client:
                await file_client.upload_data(data=item, overwrite=True)
            async with dir_client.get_file_client("foo2") as file_client:
                await file_client.upload_data(data=item, overwrite=True)

if __name__ == "__main__":
    asyncio.run(main())
```

**Expected behavior**
The above would run without problems.

**Comment by jalauzon-msft** [comment]
Hi @oliverw1 Oliver, this behavior is expected due to how we handle the HTTP pipeline and connection pooling in the SDK. 

When you open a client, we create an internal HTTP pipeline with a connection pool. Then "navigation methods" like `get_file_client` and `get_directory_client` pass that pipeline and connection pool into the new client rather than creating a whole new one. This allows client to share a connection pool which reduces the overhead of opening new HTTP connections for each new client. This helps in situations where you are maybe listing all files in a large directory and then performing an operation on each. Each FileClient you create does not need to spin up a new connection pool and can use the shared pool.

Consequently, this means when you close a client, we call close on the underlying connection pool, causing any client using that shared connection pool to stop working. The recommendation is to only use your top-level client in a context manager and there is no need to close clients created from navigation methods. So, you example would become like below. The connection pool (the main resource that needs to be closed) is handled by your top level `DataLakeDirectoryClient`

```py
async def main():
    async with DataLakeDirectoryClient.from_connection_string(
        conn_str, file_system_name="foo", directory_name="bar"
    ) as dir_client:
        file_client = dir_client.get_file_client("foo1")
        await file_client.upload_data(data=item, overwrite=True)
        file_client = dir_client.get_file_client("foo2")
        await file_client.upload_data(data=item, overwrite=True)
```

Hopefully that helps to explain this behavior. Thanks!

**Comment by oliverw1** [comment]
Thanks for the explanation, Jacob. It sounds like the same process that is used in the non async version, though [over there](https://github.com/Azure/azure-sdk-for-python/issues/14497) a fix was created. In fact, [one of the tests there](https://github.com/Azure/azure-sdk-for-python/pull/14910/files#diff-53ffdec51ac27133ad7619bcd356ca90a9c596e1fc7a7a88ec183b0f4835eb2c) show this behavior was allowed in the async version. And [it's still there](https://github.com/Azure/azure-sdk-for-python/blob/dffd1a608c295d843ef2035f27591d92eab12a24/sdk/storage/azure-storage-file-datalake/tests/test_file_system_async.py#L1153), so I am puzzled this would not work.

I'm okay using the workaround, just puzzled by that test.

**Comment by jalauzon-msft** [comment]
Hi again @oliverw1 Oliver, thanks for pointing out that test. It turns out the fix that was made for #14497 that you pointed out, was actually a way to prevent this type of issue while maintaining the same connection pool. However, the fix actually had a bug where this wasn't properly working specifically for `DataLakeDirectoryClient`. This is why your example was not working but our test was, our test used `FileSystemClient` as the parent client which did not have a bug.

We will create a PR to fix this `DataLakeDirectoryClient` soon. Thanks for your help finding this issue!
